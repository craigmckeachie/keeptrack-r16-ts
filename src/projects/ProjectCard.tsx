import { Project } from './Project';
import React from 'react';
import { Link } from 'react-router-dom';

interface ProjectCardProps {
  project: Project;
  onEdit: (project: Project) => void;
}

function formatDescription(description: string): string {
  return description.substring(0, 60) + '...';
}

function ProjectCard(props: ProjectCardProps) {
  const { project, onEdit } = props;
  const handleEditClick = (projectBeingEdited: Project) => {
    onEdit(projectBeingEdited);
  };
  return (
    <div className="card">
      <img src={project.imageUrl} alt={project.name} />
      <section className="section dark">
        <Link to={'/projects/' + project.id}>
          <h5 className="strong">
            <strong>{project.name}</strong>
          </h5>
          <p>{formatDescription(project.description)}</p>
        </Link>
        <p>Budget : {project.budget.toLocaleString()}</p>
        <button
          className="bordered"
          onClick={() => {
            handleEditClick(project);
          }}
        >
          <span className="icon-edit " />
          Edit
        </button>
      </section>
    </div>
  );
}

export default ProjectCard;
