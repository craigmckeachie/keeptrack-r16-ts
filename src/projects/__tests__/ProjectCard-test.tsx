import { ShallowWrapper, shallow } from 'enzyme';
import ProjectCard from '../ProjectCard';
import React from 'react';
import { Project } from '../Project';

describe('<ProjectCard />', () => {
  let wrapper: ShallowWrapper;
  let project: Project;
  let handleEdit: jest.Mock;

  beforeEach(() => {
    project = new Project({
      id: 1,
      name: 'Mission Impossible',
      description: 'This is really difficult',
      budget: 100
    });
    handleEdit = jest.fn();
    wrapper = shallow(<ProjectCard project={project} onEdit={handleEdit} />);
  });

  test('renders without crashing', () => {
    expect(wrapper).toBeDefined();
  });

  test('renders project prop properly', () => {
    const h5 = wrapper.find('h5');
    // const editButton = wrapper.find('button');
    const descriptionParagraph = wrapper.find('p').first();
    const budgetParagraph = wrapper
      .find('p')
      .filterWhere(n => n.text().startsWith('Budget :'));

    expect(h5.text()).toContain(project.name);
    //how to replace formatDescription, watch first video of React Testing Recipes or Cookbook
    //how to better test budget.toLocaleString
    // see https://jestjs.io/docs/en/mock-functions#mocking-modules
    expect(descriptionParagraph.text()).toEqual(project.description + '...');
    expect(budgetParagraph.text()).toContain('100');
  });

  test('handler called when edit clicked', () => {
    const editButton = wrapper.find('button');
    editButton.simulate('click');
    expect(handleEdit).toBeCalledWith(project);
  });

  //favor shallow snapshots so it is easier to see regressions
  test('snapshot', () => {
    // const tree = renderer
    //   .create(<ProjectCard project={project} onEdit={handleEdit} />)
    //   .toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});
