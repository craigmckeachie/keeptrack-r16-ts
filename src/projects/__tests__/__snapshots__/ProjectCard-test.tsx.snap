// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectCard /> snapshot 1`] = `ShallowWrapper {}`;
